kubernetes/resource_kubernetes_env.go

```diff
@@ -425,37 +463,75 @@ func resourceKubernetesEnvUpdate(ctx context.Context, d *schema.ResourceData, m
 		env = []map[string]interface{}{}
 	}
 
-	spec := map[string]interface{}{
-		"template": map[string]interface{}{
-			"spec": map[string]interface{}{
-				"containers": []interface{}{
-					map[string]interface{}{
-						"name": d.Get("container").(string),
-						"env":  env,
+	var spec map[string]interface{}
+	if container := d.Get("container").(string); container != "" {
+		spec = map[string]interface{}{
+			"template": map[string]interface{}{
+				"spec": map[string]interface{}{
+					"containers": []interface{}{
+						map[string]interface{}{
+							"name": container,
+							"env":  env,
+						},
 					},
 				},
 			},
-		},
-	}
-	if kind == "CronJob" {
-		// patch for CronJob
+		}
+
+		if kind == "CronJob" {
+			// patch for CronJob
+			spec = map[string]interface{}{
+				"jobTemplate": map[string]interface{}{
+					"spec": map[string]interface{}{
+						"template": map[string]interface{}{
+							"spec": map[string]interface{}{
+								"containers": []interface{}{
+									map[string]interface{}{
+										"name": container,
+										"env":  env,
+									},
+								},
+							},
+						},
+					},
+				},
+			}
+		}
+	} else {
 		spec = map[string]interface{}{
-			"jobTemplate": map[string]interface{}{
+			"template": map[string]interface{}{
 				"spec": map[string]interface{}{
-					"template": map[string]interface{}{
-						"spec": map[string]interface{}{
-							"containers": []interface{}{
-								map[string]interface{}{
-									"name": d.Get("container").(string),
-									"env":  env,
+					"initContainers": []interface{}{
+						map[string]interface{}{
+							"name": d.Get("init_container").(string),
+							"env":  env,
+						},
+					},
+				},
+			},
+		}
+
+		if kind == "CronJob" {
+			// patch for CronJob
+			spec = map[string]interface{}{
+				"jobTemplate": map[string]interface{}{
+					"spec": map[string]interface{}{
+						"template": map[string]interface{}{
+							"spec": map[string]interface{}{
+								"initContainers": []interface{}{
+									map[string]interface{}{
+										"name": d.Get("init_container").(string),
+										"env":  env,
+									},
```

Comment:
Similar to my other comment, there's a fair bit of similar code being repeated here. Maybe there's a way you could use a variable to switch between `initContainers` and `containers` since they have the same structure.

Reply:
made the map of name and env a variable, not sure if we can make the whole jobTemplate/spec/template/spec structure a variable but I may have just missed it.